src/vanilla.jl

```diff
@@ -244,106 +262,131 @@ function plan!(planner::AbstractMCTSPlanner, s)
     return tree
 end
 
+
 function build_tree(planner::AbstractMCTSPlanner, s)
     n_iterations = planner.solver.n_iterations
     depth = planner.solver.depth
 
     if planner.solver.reuse_tree
         tree = planner.tree
     else
-        tree = MCTSTree{statetype(planner.mdp), actiontype(planner.mdp)}(n_iterations)
+        tree = MCTSTree{statetype(planner.mdp), actiontype(planner.mdp)}(s)
     end
 
-    sid = get(tree.state_map, s, 0)
-    if sid == 0
+    root = get(tree.states, s, nothing)
+    if root == nothing
         root = insert_node!(tree, planner, s)
-    else
-        root = StateNode(tree, sid)
     end
 
-    start_us = CPUtime_us()
-    # build the tree
-    for n = 1:n_iterations
-        simulate(planner, root, depth)
-        if CPUtime_us() - start_us >= planner.solver.max_time * 1e6
-            break
+    timeout_us = CPUtime_us() + planner.solver.max_time * 1e6
+    # Run simulation in a sequential manner in case of single thread. The
+    # Channel approach below seems more efficient only if multiple threads are
+    # used. This is presumably due that in case of single thread a lot of time
+    # can be wasted context switching between the producer and consumer.
+    if Threads.nthreads() == 1
+        for n in 1:n_iterations
+            simulate(planner, tree, root, depth)
+            CPUtime_us() > timeout_us && break
+        end
+        return tree
+    end
+
+    # Use a Channel to implement a producer-consumer type interaction. This
+    # seems quite crucial for performance in case of multiple threads.
+    # TODO(kykim): See if the two cases can be more concisely combined.
+    sim_channel = Channel{Task}(min(1000, n_iterations)) do channel
+        for n in 1:n_iterations
+            put!(channel, Threads.@spawn simulate(planner, tree, root, depth, timeout_us))
+        end
+    end
+    for sim_task in sim_channel
+        CPUtime_us() > timeout_us && break
```

Comment:
This should not be CPUtime anymore because we are multithreading - see how @WhiffleFish did it here: https://github.com/JuliaPOMDP/POMCPOW.jl/pull/32

Reply:
Thanks! That looks like a nice find